networkit/cpp/randomization/DegreePreservingShuffle.cpp

```diff
@@ -0,0 +1,169 @@
+#include <omp.h>
+
+#include <algorithm>
+#include <vector>
+
+#include <networkit/Globals.hpp>
+#include <networkit/auxiliary/Parallel.hpp>
+#include <networkit/auxiliary/Random.hpp>
+#include <networkit/auxiliary/SignalHandling.hpp>
+#include <networkit/graph/GraphTools.hpp>
+#include <networkit/randomization/DegreePreservingShuffle.hpp>
+
+namespace NetworKit {
+
+// The algorithms effectively computes for each degree (in the undirected case it's a scalar, in the directed case it's a scalar pair)
+// the set of nodes which have this degree and randomly reassigns ids *within* these sets. Since directed and undirected degrees
+// have a different signature, the algorithm itself is templated and compares degrees only using the < and the == operator.
+// Those implementation details are hidden in the DegreePreservingShuffleDetails namespace:
+namespace DegreePreservingShuffleDetails {
+struct DirectedDegree {
+    count in;
+    count out;
+    bool operator< (const DirectedDegree& o) const {return std::tie(in, out) <  std::tie(o.in, o.out);}
+    bool operator==(const DirectedDegree& o) const {return std::tie(in, out) == std::tie(o.in, o.out);}
+};
+
+template <typename DegreeT>
+struct NodeDegree {
+    node id;
+    union {
+        DegreeT degree;
+        node alt_id;
+    };
+
+    NodeDegree() {}
+};
+
+using UndirectedNodeDegree = NodeDegree<count>;
+using DirectedNodeDegree = NodeDegree<DirectedDegree>;
+
+template <typename DegreeT>
+static std::vector<index> computePermutation(std::vector<NodeDegree<DegreeT>>& node_degrees) {
+    const auto n = node_degrees.size();
+    using node_degree_t = NodeDegree<DegreeT>;
+
+    // Allocate memory for the permutation
+    std::vector<index> permutation(n);
+    #ifndef NDEBUG
+    std::fill_n(permutation.begin(), n, std::numeric_limits<node>::max());
+    #endif
+
+    // We sort by degree and thereby move nodes of equal degrees into consecutive groups
+    Aux::Parallel::sort(begin(node_degrees), end(node_degrees),
+                        [] (const node_degree_t a, const node_degree_t b) {return a.degree < b.degree;});
+
+    // Now shuffle each group
+    count num_changes = 0;
+    const auto requested_threads = std::max(1, std::min<int>(omp_get_max_threads(), n / 50000));
+    #pragma omp parallel num_threads(requested_threads) reduction(+: num_changes)
+    {
+        const auto tid = omp_get_thread_num();
+        const auto nthreads = omp_get_num_threads();
+
+        const auto chunk_size = (n + nthreads - 1) / nthreads;
+        auto chunk_begin = node_degrees.begin() + std::min<size_t>(chunk_size * tid, n);
+        auto chunk_end   = node_degrees.begin() + std::min<size_t>(chunk_size * (tid + 1), n);
+
+        using iter_t = decltype(chunk_begin);
```

Comment:
Same issue. But here the concrete type is too long to type. Maybe use `DegreeIterator`. Unfortunately, we're not in C++14 yet, where we could just make the lambda argument a generic `auto`.